backend/src/controller/chatController.ts

```diff
@@ -160,90 +220,119 @@ async function handleChatToGPT(req: OpenAiChatRequest, res: Response) {
 				message,
-				chatHistoryBefore,
-				chatResponse,
+				initChatResponse,
 				currentLevel,
-				chatModel
+				chatModel,
+				currentChatHistory,
+				defences
 			);
 		}
 	} catch (error) {
 		const errorMessage =
 			error instanceof Error ? error.message : 'Failed to get chatGPT reply';
-		handleErrorGettingReply(req, res, currentLevel, chatResponse, errorMessage);
+		req.session.levelState[currentLevel].chatHistory = addErrorToChatHistory(
+			currentChatHistory,
+			errorMessage
+		);
+		handleChatError(res, initChatResponse, errorMessage, 500);
 		return;
 	}
 
-	if (chatResponse.defenceReport.isBlocked) {
+	let updatedChatHistory = levelResult.chatHistory;
+	totalSentEmails.push(...levelResult.sentEmails);
+
+	// update chat response
+	const updatedChatResponse: ChatHttpResponse = {
+		...initChatResponse,
+		reply: levelResult.chatResponse.reply,
+		wonLevel: levelResult.chatResponse.wonLevel,
+		openAIErrorMessage: levelResult.chatResponse.openAIErrorMessage,
+		sentEmails: levelResult.sentEmails,
+		defenceReport: levelResult.chatResponse.defenceReport,
+		transformedMessage: levelResult.chatResponse.transformedMessage,
+	};
+
+	if (updatedChatResponse.defenceReport.isBlocked) {
 		// chatReponse.reply is empty if blocked
-		req.session.levelState[currentLevel].chatHistory.push({
+		updatedChatHistory = pushMessageToHistory(updatedChatHistory, {
 			completion: null,
 			chatMessageType: CHAT_MESSAGE_TYPE.BOT_BLOCKED,
-			infoMessage: chatResponse.defenceReport.blockedReason,
+			infoMessage: updatedChatResponse.defenceReport.blockedReason,
 		});
-	}
-	// more error handling
-	else if (chatResponse.openAIErrorMessage) {
-		handleErrorGettingReply(
-			req,
-			res,
-			currentLevel,
-			chatResponse,
-			simplifyOpenAIErrorMessage(chatResponse.openAIErrorMessage)
+	} else if (updatedChatResponse.openAIErrorMessage) {
+		const errorMsg = simplifyOpenAIErrorMessage(
+			updatedChatResponse.openAIErrorMessage
 		);
+		req.session.levelState[currentLevel].chatHistory = addErrorToChatHistory(
+			updatedChatHistory,
+			errorMsg
+		);
+		handleChatError(res, updatedChatResponse, errorMsg, 500);
 		return;
-	} else if (!chatResponse.reply) {
-		handleErrorGettingReply(
-			req,
-			res,
-			currentLevel,
-			chatResponse,
-			'Failed to get chatGPT reply'
+	} else if (!updatedChatResponse.reply) {
+		const errorMsg = 'Failed to get chatGPT reply';
+		req.session.levelState[currentLevel].chatHistory = addErrorToChatHistory(
+			updatedChatHistory,
+			errorMsg
 		);
+		handleChatError(res, updatedChatResponse, errorMsg, 500);
 		return;
+	} else {
+		// add bot message to chat history
+		updatedChatHistory = pushMessageToHistory(updatedChatHistory, {
+			completion: {
+				role: 'assistant',
+				content: updatedChatResponse.reply,
+			},
+			chatMessageType: CHAT_MESSAGE_TYPE.BOT,
+		});
 	}
 
-	// update sent emails
-	chatResponse.sentEmails =
-		req.session.levelState[currentLevel].sentEmails.slice(numSentEmails);
```

Comment:
What the heck is this line meant to do. And do we still want to be doing that?

Reply:
Ah okay, array.slice(index) will return a sub-array from the index, to the end of the array. In practise, this line extracts only those new emails that have been sent as a result of this particular chat message so that they can be returned without returning the entire list of emails. We are still doing this (and now it's much easier to read!), so no regression. ✅

Reply:
Haha, you explained it so much better than me 😅